### PR TITLE
Preserve type Annotations

### PR DESCRIPTION
**Pull request by paulo-raca**
## Description

Python's `typing.Annotated` is a handy way of appending misc info into a field, argument, etc.

In fact, Strawberry uses it in a few places, like `strawberry.Private` and `strawberry.argument`.

However Strawberry discards this information when processing types, which prevents other types of annotations from being used.

This PR introduces `StrawberryAnnotated`, a StrawberryContainer to store this kind of information.

(using `typing.Annotated` would probably be ideal, however I can only supports python types, not StrawberryType instances)

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

(I think this is an internal aspect that doesn't need to be documented on the public docs, let me know what do you think)

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

Add StrawberryAnnotated, a container type to preserve `typing.Annotated` data when converting types to StrawberryTypes


---
Here's the preview release card for twitter:




Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Paulo Costa for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: RELEASE.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Release type: minor
+
+Add StrawberryAnnotated, a container type to preserve `typing.Annotated` data when converting types to StrawberryTypes
```
Comment (estyxx) [inline RELEASE.md:3]:
```suggestion
This release adds support for `typing.Annotated` so that is preserved  when converting types to StrawberryTypes
```
I would rephrase a little to emphasize the functionality we are adding rather than what changed internally :)

Comment (paulo-raca) [reply RELEASE.md]: Sounds good, thank you

File: strawberry/type.py [hunk 2]
```diff
@@ -51,18 +53,25 @@ class StrawberryContainer(StrawberryType):
...
         self.of_type = of_type
 
-    def __hash__(self) -> int:
-        return hash((self.__class__, self.of_type))
-
     def __eq__(self, other: object) -> bool:
         if isinstance(other, StrawberryType):
-            if isinstance(other, StrawberryContainer):
+            if type(self) == type(other):
+                other = cast(StrawberryContainer, other)
```
Comment (BryceBeagle) [inline strawberry/type.py:59]: Not sure I understand why two objects being of the same type means that one of them is a container type. Can you elaborate here?

File: strawberry/type.py [hunk 3]
```diff
@@ -51,18 +53,25 @@ class StrawberryContainer(StrawberryType):
...
                 return self.of_type == other.of_type
             else:
                 return False
 
         return super().__eq__(other)
 
+    def __hash__(self) -> int:
+        return hash((type(self), self.of_type))
```
Comment (BryceBeagle) [inline strawberry/type.py:67]: Can we get test coverage here?

File: strawberry/type.py [hunk 4]
```diff
@@ -118,6 +127,66 @@ class StrawberryOptional(StrawberryContainer):
     ...
 
 
+class StrawberryAnnotated(StrawberryContainer):
+    """
+    Equivalent to typing.Annotated for strawberry types
+    """
```
Comment (BryceBeagle) [inline strawberry/type.py:133]:
```suggestion
    """Equivalent to typing.Annotated for Strawberry types"""
```
nit

File: strawberry/type.py [hunk 5]
```diff
@@ -118,6 +127,66 @@ class StrawberryOptional(StrawberryContainer):
...
+
+    def __init__(self, of_type: Union[StrawberryType, type], *args):
+        of_type, base_args = StrawberryAnnotated.get_type_and_args(of_type)
+        super().__init__(of_type)
+        self.args = base_args + list(args)
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, StrawberryType):
+            if type(self) == type(other):
```
Comment (BryceBeagle) [inline strawberry/type.py:142]: Why not `isinstance(other, type(self))`?
This would also be resilient if we ever have a subclass for this type.

File: strawberry/type.py [hunk 6]
```diff
@@ -118,6 +127,66 @@ class StrawberryOptional(StrawberryContainer):
...
+                other = cast(StrawberryAnnotated, other)
+                return (self.of_type, self.args) == (other.of_type, other.args)
+            else:
+                return False
+
+        return super().__eq__(other)
```
Comment (BryceBeagle) [inline strawberry/type.py:148]: Missing a bit of coverage in this file. Could you add some quick tests?

File: strawberry/type.py [hunk 7]
```diff
@@ -118,6 +127,66 @@ class StrawberryOptional(StrawberryContainer):
...
+    def __hash__(self) -> int:
+        return hash((type(self), self.of_type, self.args))
+
+    def __str__(self) -> str:
+        args = ", ".join(map(str, [self.of_type] + self.args))
+        return f"{type(self).__name__}({args})"
+
+    def __repr__(self) -> str:
+        args = ", ".join(map(repr, [self.of_type] + self.args))
+        return f"{type(self).__name__}({args})"
```
Comment (BryceBeagle) [inline strawberry/type.py:159]: Why define both `__str__` and `__repr__` identically? If `__repr__` is defined, but `__str__` is not, `repr()` will use `__str__`. 

(https://stackoverflow.com/a/2626364)

File: strawberry/type.py [hunk 8]
```diff
@@ -118,6 +127,66 @@ class StrawberryOptional(StrawberryContainer):
...
+
+    def copy_with(
+        self, type_var_map: Mapping[TypeVar, Union[StrawberryType, type]]
+    ) -> StrawberryType:
```
Comment (BryceBeagle) [inline strawberry/type.py:163]: I think the return type here can be narrowed to `StrawberryAnnotated`?

File: strawberry/type.py [hunk 9]
```diff
@@ -118,6 +127,66 @@ class StrawberryOptional(StrawberryContainer):
...
+        # Parent implementation handles the nested type,
+        # but discards the arguments
+        ret = cast(StrawberryAnnotated, super().copy_with(type_var_map))
+        return StrawberryAnnotated(ret.of_type, self.args)
```
Comment (BryceBeagle) [inline strawberry/type.py:167]: Why make a copy, and then another instance?

File: strawberry/type.py [hunk 10]
```diff
@@ -118,6 +127,66 @@ class StrawberryOptional(StrawberryContainer):
...
+        while True:
+            if isinstance(type, StrawberryAnnotated):
+                args = type.args + args
+                type = type.of_type
+            elif get_origin(type) is Annotated:
+                base_args = get_args(type)
+                type = base_args[0]
+                args = list(base_args[1:]) + args
+            else:
+                return type, args
```
Comment (BryceBeagle) [inline strawberry/type.py:187]:
```suggestion
    def get_type_and_args(
        type_: Union[StrawberryType, type]
    ) -> Tuple[Union[StrawberryType, type], List[Any]]:
        """
        Splits a possibly-annotated type in the actual type and the list of annotations
        Supports both StrawberryAnnotated and typing.Annotated
        """
        args: List[Any] = []
        while True:
            if isinstance(type_, StrawberryAnnotated):
                args = type_.args + args
                type = type_.of_type
            elif get_origin(type_) is Annotated:
                base_args = get_args(type_)
                type_ = base_args[0]
                args = list(base_args[1:]) + args
            else:
                return type_, args
```

**Review by BryceBeagle — CHANGES_REQUESTED** [review]
This is a great contribution! I've left some requests, but hopefully we can get this merged in soon :+1:

**Comment by paulo-raca** [comment]
I've addressed the comments and fixed the coverage.

I also gave some attention/testing to StrawberryTypes.hash/equals/repr

**Comment by paulo-raca** [comment]
I have rebased this change on top of the current strawberry version (No major conflicts :joy: )

**Review by paulo-raca — COMMENTED** [review]
@BryceBeagle, any thoughts on this?

Any chance this is getting merged? Or should I drop it instead?
